### DISCUSSION TITLE
Installing cedar-detect

### DISCUSSION THREAD

**Discussion by AstroKeith** [smroid/cedar-detect#1] [General]
HI,
I have Tetra3 running OK (On both a Mac mini and Raspberry Pi5 running Bookworm (Debian 12)
I want to try cedar-detect but am having trouble installing. (I'm not an advanced user!)
I've successfully installed gprc and gprc_tools. But then stuck.
Could you provide some guidance on the rest of the install process.
Many thanks 
Keith

**Comment by smroid** [comment]
Hi Keith,

I assume you have an existing system that's using Tetra3, and you want to try out cedar-detect as a replacement for Tetra3's star detection and centroiding. My response is based on this assumption; please let me know if you're headed in a different direction.

I'll broaden the scope by suggesting that you should also migrate to cedar-solve as a replacement for Tetra3. cedar-solve is a Tetra3 fork that greatly improves Tetra3's average and worst case performance. If you do migrate, you'll want to look into replacing your pattern database using cedar-solve as it creates much more efficient (small) pattern databases.

https://github.com/smroid/cedar-solve/blob/master/migrating.rst outlines a two step approach for this migration. The first is to switch from Tetra3's plate solving to cedar-solve's plate solving. The second step is to swap in cedar-detect to replace Tetra3 (and cedar-solve) star detection and centroiding.

Let me know if this helps,

-Steven

**Reply by AstroKeith** [reply]
Hi, not working yet.
I managed to follow your excellent migration instructions. (had to 'sudo apt-get install protobuf-compiler') along the way.
I've updated my PYTHONPATH and changed my calling script.
It finds the new tetra3, but in line 135 fails...
![Screenshot 2024-07-17 at 14 46 43](https://github.com/user-attachments/assets/968ef3ca-0fd3-4260-88d2-82b7e886f0c0)
Any suggestions?

**Reply by AstroKeith** [reply]
Update: It seems this may all be due to the folder and script having the same name tetra3. BTW I'm running Python 3.11

**Comment by AstroKeith** [comment]
Thanks - thats exactly what I want to do. 
In the next day or two I'll try your migration steps. Stand by!

Keith

**Comment by smroid** [comment]
The cedar-solve component was recently reorganized by a contributor to have
a more formal packaging and installation system. Try following the
instructions at
https://github.com/smroid/cedar-solve/blob/master/docs/installation.rst;
the issue might be the need to define/use a suitable venv for cedar-solve
so it can find its own dependencies.

Good luck!
-Steven

**Reply by AstroKeith** [reply]
Thanks. The pip install into a venv is the way ahead.
That helped quite a lot but a couple of issues so far....

in cedar_detect_pb2_grpc.py, line 5 it should be 'from tetra3 import cedar_detect_pb2 as cedar__detect__pb2'
The pip install doesnt make the needed cedar-detect-server binaries. (Luckily I had kept the one created from my previous install)

**Reply by AstroKeith** [reply]
Update:
With the changes I made in my last message, I have it running on a Pi5 with 64bit Bookworm OS (Debian 12)
On that platform it is indeed between x2 and x3 faster than baseline Tetra3. Nice! Good job. Importantly the odd fail I got with Tetra3 has gone, although with foreground obstructions solve time goes into seconds. I'll try tweaking some of the parameters next.
A am occasionally getting warnings and errors about shared memory issues from cedar-detect-server. I'm seeing these when I have to abort the script for some reason and then run again. They clear if I reboot the Pi. Is there some clean-up process needed?

Question:! The binaries needed, that currently have to be created through Rust - could these just be downloadable?

**Reply by smroid** [reply]
I'm trying to repro and fix the issue with cedar-detect-server being left in a bad state.

When you say "when I have to abort the script", what script are you referring to and how are you aborting it?

When I run the test_tetra3.py program in cedar-solve/examples, if I kill it with ^C the cedar-detect-server process goes away. However when I kill the python process (the one running test_tetra3.py) with kill -9, the cedar-detect-server process lingers. 

Please let me know if you recall any details on what you were doing/seeing. In the meantime I'll try to figure out if I can cause cedar-detect-server to go away even if the parent is killed with kill -9.

**Reply by smroid** [reply]
I've pushed a fix to the cedar-detect repo. You can git pull it and then 'cargo build --release' to build the cedar-detect-server binary with the fix.

Then, copy the binary to cedar-server/tetra3/bin; alternately you can git pull the latest cedar-solve which installs a symlink in cedar-server/tetra3/bin.

Let me know if this helps!

**Reply by AstroKeith** [reply]
Thanks Steven,
I'm still getting the issue. Screen grab attached,
It seems to be triggered when I invoke the arg return_visual=True, and then use PIL to show it. Just invoking the arg doesnt cause the issue.
You can see that even the error print out has been truncated and the Pi5 is left hanging. It seems the script I has called ran and closed albeit with warnings, then the command line shows resource_tracker.py hanging.
I wonder if this is related to the old PILLOW issue I reported earlier. the pip install of cedar-solve failed as it tries to install a very old version of PILLOW. I eventually got over this by manual intervention but now my venv is running PIL 8.4.0. Having installed via pip I cant upgrade PILLOW.
My final application won't use PIL to show the visual, but while I am refining parameters and assessing cameras, I find it a useful tool.
On the PiZero, clearly I'm not trying to invoke return_visual-True, so I cannot reproduce that error, but I still get a load of gprc warnings. However it runs. 2nd screen grab. I didnt use pip for this install. 
This weekend I'll do a fresh install on both the Pi5 and PiZero, not using pip.

I'm using 64 bit Bookworm (Debian 12) throughout, which includes python 3.11

**Reply by smroid** [reply]
I don't see the screen grabs... I've not used the return_visual=True arg, I'm trying it out now.

I was planning on removing the return_visual arg from the solve_from_centroids() call, and instead have the caller decorate the image based on info from return_matches=True; see examples/test_tetra3.py for how I'm using that.

**Reply by AstroKeith** [reply]
Oops!
![Screenshot 2024-07-20 at 08 12 20](https://github.com/user-attachments/assets/2a633693-baae-45d1-a439-880adb91836d)
![Screenshot 2024-07-20 at 08 49 29](https://github.com/user-attachments/assets/832f1a9b-287d-41ed-b9a0-06a7ece3e16f)

**Reply by smroid** [reply]
For the second screenshot (white background), I see those as well. Not sure what they're about but seem benign.

**Reply by AstroKeith** [reply]
I only get those on the PiZero.
I get a nice clean report on the Pi5. Supposedly the same basic OS but clearly very different hardware. Screen shot of one of my test runs.
![Screenshot 2024-07-20 at 17 48 43](https://github.com/user-attachments/assets/7288dbd0-cd82-4fa0-9b55-cf4e66558661)

**Reply by smroid** [reply]
Nice numbers!

I tried return_visual=True, and rather than trying to display the image I instead write it to file and inspect offline. It all seems to work fine.

Can you send me the code fragment you're using to display the image using PIL?

**Reply by AstroKeith** [reply]
Attached the python script I use to evaluate options against images. A lot of it is about sorting files and loading appropriate databases. The offending im.show is commented out near the bottom.
[test_tetra3_cedar.py.zip](https://github.com/user-attachments/files/16323994/test_tetra3_cedar.py.zip)
I've done a fresh build for the PiZero (trying to get my own efinder install.sh correct!).
On running it threw up an error in line 97 or cedar_detect_client.py at binning=binning. This is a new arg I think, so did I get an interim build? I tried adding a binning=1, arg to the call, but that didnt work.
The gprc warnings I get on the PiZero (previous white background screenshot) only seem to pop up on the first centroid call. Subsequent ones are clear.
One of eFinder functions is a focus mode. This uses the centroid-detect to find the brightest potential star and loop through the capture/centroid-detect process, displaying an image patch, number of centroids and a point spread function through the star. (photo below). I'm finding cedar ignores stars until they are quite well focused, so I am reverting to tetra3 for this focus function. Is this inherent in your fast detection process, or can I pass some arguments to force it to include out of focus stars?
![IMG_6412](https://github.com/user-attachments/assets/8aae4cc0-e528-4211-8446-0932245f1d0d)
But overall things going well. I have updated a version of the of whole 'eFinder Lite' to use cedar and it's running smoothly. No obvious memory issues. First solve takes 1 second, but subsequent are 0.1 to 0.2s. Clearly the OS is managing memory quite well.

**Reply by smroid** [reply]
> Attached the python script I use to evaluate options against images. A lot of it is about sorting files and loading appropriate databases. The offending im.show is commented out near the bottom. 

Thanks. I'll try this out.

> I've done a fresh build for the PiZero (trying to get my own efinder install.sh correct!). On running it threw up an error in line 97 or cedar_detect_client.py at binning=binning. This is a new arg I think, so did I get an interim build? I tried adding a binning=1, arg to the call, but that didnt work.

Apologies-- I updated cedar_detect.proto but neglected to push the updated files cedar_detect_pb2.py and cedar_detect_pb2.pyi in cedar-solve/tetra3 directory. You can run 'python scripts/compile_proto.py' from the top-level cedar-solve directory to update these.

The new 'binning' parameter controls whether 2x or 4x binning is used. Legal values are None, 2, or 4. See cedar_detect.proto for doc.

> The gprc warnings I get on the PiZero (previous white background screenshot) only seem to pop up on the first centroid call. Subsequent ones are clear. 

I've been seeing these as well, they seem benign. In any case I've found a solution-- in cedar-solve/tetra3/cedar_detect_client.py, find the call ...ExtractCentroids(req) and change it to ...ExtractCentroids(req, wait_for_ready=True).

> One of eFinder functions is a focus mode. This uses the centroid-detect to find the brightest potential star and loop through the capture/centroid-detect process, displaying an image patch, number of centroids and a point spread function through the star. (photo below). I'm finding cedar ignores stars until they are quite well focused, so I am reverting to tetra3 for this focus function. Is this inherent in your fast detection process, or can I pass some arguments to force it to include out of focus stars?

You're correct, Cedar-detect requires stars to be well focused or nearly so. Cedar Aim also uses the detected star count as a user cue but it only works when the focus is already pretty good; it is useful for finding the best focus. For out-of-focus conditions Cedar Aim uses other techniques to provide focus assistance.

> But overall things going well. I have updated a version of the of whole 'eFinder Lite' to use cedar and it's running smoothly. No obvious memory issues. First solve takes 1 second, but subsequent are 0.1 to 0.2s. Clearly the OS is managing memory quite well.

Nice!

**Reply by smroid** [reply]
> Attached the python script I use to evaluate options against images. A lot of it is about sorting files and loading appropriate databases. The offending im.show is commented out near the bottom.

I tried this but it has no effect, probably because my Rpi setup is headless.

**Reply by AstroKeith** [reply]
Thanks for all this.
The ExtractCentroids(req, wait.... fix didnt work for me. But as you say it seems benign and only appears on first call.

**Comment by smroid** [comment]
Glad to hear you have it running!

Regarding foreground obstructions: within reason, cedar-solve should be fine with some foreground obstructions, though more obstructions will make things worse by blocking stars (by contrast, the old Tetra3 star detection code would detect tons of false stars on any illuminated foreground objects, guaranteeing a solve fail). Can you send me some examples of foreground obstruction images that take a long time to solve, but do succeed? I have what might be a simple fix for making those solve faster.

Regarding cedar-detect-server errors after aborting the script. Looking at cedar-solve/tetra3/cedar_detect_client.py, I think what's happening is that __init__() spawns the cedar-detect-server subprocess, and then CedarDetectClient operations talk to it.

The __del__() function kills the cedar-detect-server-subprocess, so when the calling program terminates, the cedar-detect-server subprocess is cleaned up. But I think what might be happening is if you abort the script (by ^C, right?), it won't execute __del__().

Perhaps CedarDetectClient could be improved to define a signal handler to trap ^C and ensure that __del__() is executed. Or maybe we need to put that up in the __main__ portion of whatever is calling CedarDetectClient.

Regarding posting Rust binaries for people to download-- I do provide a pre-built sdcard image which boots the Raspberry Pi right into Cedar, no need to download or build anything. At the other extreme, people can clone the repos and build everything themselves. Given these two options I don't think providing pre-built Rust binaries is adding much.

**Reply by AstroKeith** [reply]
Attached is a less than perfect image for you. In fact by tweaking the database and solve parameters I got the solve time down dramatically, although still x4 'normal'. I'm learning a lot! 

One factor is the field of view. I read somewhere that it is horizontal. It seems not to be and is it the same definition when generating databases and entering the estimate for the solver? I changed to using the height of the image (ie the minimum dimension) and it seems to be better. Perhaps this could be clarified in the --help?

About the errors. If I run a simple script performing detect and then solve - it does seems OK. A new error though -  If I add showing the visual image with markers, I now get memory leak warnings, and the script doesnt exit. I attach a screen shot of the shell window.

Understand about the binaries. My next step is to port cedar onto a Pi Zero 2 W. I have Tetra3 running OK, albeit slow for some images. I'll be using the same OS as the Pi5 (Debian 12) but the cutdown version (eg no desktop). The challenge is the small RAM.

![rpi_25_1_25_140724_221421](https://github.com/user-attachments/assets/3ee3bdb2-945f-4131-9c00-bb08dc6df20a)
![Screenshot 2024-07-18 at 08 14 13](https://github.com/user-attachments/assets/4bf1bb3a-b0d5-4de6-9073-98a7a22df638)

**Reply by smroid** [reply]
On this image with a small illuminated obstruction, I am also showing elevated solve times. This isn't because of the obstruction (it is small and cedar-detect does not see false stars in it, unlike Tetra3's star detection which "sees" many stars there), it is more down to the fact that some fields end up requiring more pattern searches, and this is one of the unlucky ones. There *might* be a correlation with near horizon images, due to atmospheric refraction differentially displacing star positions, but that's just a guess.

However, the call to solve_from_centroids() can be improved in two ways. The first is that once you know the HFOV, you can pass e.g. fov_estimate=14, fov_max_error=1 and that speeds up the solve time somewhat.

The other thing you can do is pass match_max_error=0.002 (instead of the default 0.005); this greatly reduces the solve time. This parameter is suitable if your camera system does not introduce significant distortions; at the 10deg FOV we typically use this is generally not a problem.

**Comment by smroid** [comment]
You mentioned that you're tweaking the database. For best results when using cedar-solve (instead of classic Tetra3), be sure to also use cedar-solve's tool to build the database. What parameters are you using to build your pattern db? What are your solve parameters? I'll play with the image you gave (thanks!) and report back with my findings and recommendations.

Tetra3's use of FOV are always horizontal (longer axis). What is the HFOV of your image?

Regarding "visual image with markers", that's using Tetra3's option to return visual info? I've not tested with that, could be "broken" in cedar-solve. Perhaps I should remove it if I'm not going to support it...

Regarding the Pi Zero 2 W-- compute-wise it is comparable to the Rpi3 and is (barely?) adequate for Cedar. However as you point out its RAM is small (half gig) and that could be a problem for cedar-solve when using a robust pattern database, especially at smaller FOV camera setups. For a low-cost processor, I'd recommend staying with the Rpi3, it has 1G RAM which should be enough.

**Comment by AstroKeith** [comment]
I'm using the cedar tool to generate databases. I like that they are smaller, plus I can blank out unwanted sky (southern). Currently for generation I'm just using max and min fov, which I set about 1/2 deg from the actual. For the actual, as I said I'm using the smaller fov dimension. a) that is what is reported in the solution, and b) it solves faster. Currently I'm assuming a 10.6 deg fov. For detect I'm using max size 10, sigma 8, use binned True. For solve basically I'm just using fov estimate (even if thats wrong!) The image I sent earlier reports as 10.6deg fov, whereas it is 14.3*10.6.

I'm consistently getting extracts in 50ms and solves in 20ms on a Pi5. No failed solves which for me is important. 

I have two applications, for one the Pi5 is OK. The other is currently a Pi Zero 2, mainly for power consumption reasons.  I have cedar working on the Pi Zero and currently getting times about x3 compared with the Pi5. This is just acceptable. gprc is throwing up warnings, but they arent fatal and I get results. I guess these times might improve after tweaking!

**Comment by AstroKeith** [comment]
I goofed somewhere!
Indeed using the width gives better results. I must have changed too many variables at the same time and come to the wrong conclusion.

On the Pi Zero, detection is 190+/-10ms and solve is about 60ms, but some big ones in there too. At least I have something to go on. I have images taken 10 seconds apart, same sky, same settings and yet solve times is x4.

An interesting result if I change use_binning to False. On the Pi5 it has no effect. On the Pi Zero, it increases solve time enormously (no effect on detect) I'm already binning in the camera.

Looking at RAM and cpu utilisation. Shows the 512MB RAM is sufficient (no swapfile being used) but the whole process is executed on one core at 100%.

**Comment by smroid** [comment]
The times you cite are all with cedar-solve using cedar-detect? The image you posted earlier is representative for these? I'll  try it out this evening and see what times I come up with.

What camera are you using? You say you're binning in the camera; cedar would prefer to see the full-res image from the camera and do any needed binning itself. This is so cedar-detect can apply its own hot pixel rejection logic.

Regarding cores-- cedar-detect and cedar-solve are single-threaded, so only one core will be used. At a higher level the cedar-server system choreographs a multi-stage pipeline of image acquisition, star detection, and plate solving, where the stages run concurrently and thus utilize multiple cores.

**Reply by AstroKeith** [reply]
Yes, they are all with cedar-detect. That image I posted is one that gives a long solve. I'll add below a more typical one.

I have two cameras I am considering. The ASI120MM produces really excellent images, but at a price. So my current work is on the Pi HQ camera, with a 25mm f1.2 lens.

If it's clear tonight I'll try to capture some test images without camera binning. The Pi HQ camera isnt great and I have found that with astrometry.net and tetra3 camera binning was almost essential!

My application doesnt call for rapid sequencing of images, except for very occasional focussing. I had guessed the cedar-server would be able to utilise more cores.

From my current stock of test images, excluding obvious bad ones, I'm getting a combined detect and solve time of circa 200ms on average.  This is pretty good (well done!) and compares well with the camera exposure time.

I know some applications are calling for continuous solving at high rate, replacing encoder systems. Mine is part of an integrated telescope control system, and is polled by the control system only to 'sync' the telescope to real sky, at specific times.
![rpi_25_1_25_140724_221308](https://github.com/user-attachments/assets/1c853d56-2008-4866-b567-c9f03d93d56d)

**Comment by smroid** [comment]
Ah, the HQ camera. Yes, cedar's preferred approach with that sensor is full-res readout, and 4x binning going into detection. See comments at https://github.com/smroid/cedar-server/blob/bbfb1140763f6553bbef7964dd91318f075dc9ad/src/bin/cedar_server.rs#L1221

I'm in the process of updating the cedar detect's call-from-python path to allow 4x binning to be specified; I'll ping you when that's ready.

So your 200ms detect/solve is on the HQ cam? That matches up with my experience with that camera.

**Comment by AstroKeith** [comment]
Thanks for the link to explanations.
Yes 200ms is with the Pi HQ camera, with a 1014x760 size call in picamera2

**Comment by smroid** [comment]
1014x760 is detecting/centroiding in 190ms on Rpi zero  and 50ms on Rpi 5? That seems high actually, did you build cedar-detect with the --release (optimized) flag?

**Reply by AstroKeith** [reply]
The Pi5 is closer to 35ms average now.

Didnt know about the --release flag. Can you point me to a description of the options, etc

**Reply by smroid** [reply]
The 35ms is for the cedar-detect, or the overall cedar-detect+solve?

The --release flag is part of the Rust toolchain, the 'cargo' tool for doing package and build management.

https://doc.rust-lang.org/cargo/commands/cargo-build.html, scroll down to --release description.

**Comment by AstroKeith** [comment]
35ms is for the detect, which is pretty much fixed. Solve varies from 10 to 50ms.
Checking my script, that time includes reading the file and converting the Pi colour image to 'L' mode. Take 15ms off!

Which leads me to another insight. Removing the disk read and image conversion from the timing on the Pi Zero reduces cedar-detect to 40ms, from 180ms. The image convert is almost 100ms!

I'm with you now on the --release, yes I did. The whole installation/build is so involved I couldn't remember all the steps I did. Is there any prospect of removing the need for Rust/gprc/etc ?

**Reply by smroid** [reply]
OK, so 20ms detect on the Rpi 5 and 40ms detect on the Rpi zero, those seem more reasonable for that image size. I suspect the average detect times on Rpi 5 are actually better than 20ms.

Re removing Rust-- cedar detect, if not written in Rust, would need to be C/C++ to achieve the performance we're seeing. I chose Rust in part because I was looking to learn Rust and the idea for Cedar came at around the same time (inspired by the Sky&Telescope writeup on Dale Eason's SkySolve last year).

Regarding gRPC: I'm doing network calls where the web app on the phone is calling out to the Rpi cedar-server. Choices there would be something JSON/RESTy or gRPC. I'm fluent in gRPC at my day job, so I went with that.

Once gRPC is in hand, we then turn to the question of interop between Rust and Python. There are (likely) interfaces to allow Rust code to be linked into Python, but I've struggled with inter-language bindings in the past and felt that gRPC has adequate inter-process performance.

So yeah, lots of moving parts and associated teething pain, for which I apologize. But the end result greatly improved performance compared to what we had before. We're now limited by the exposure time, more or less.

**Reply by AstroKeith** [reply]
Thanks for all the assistance!
match_max_error=0.002, helped a bit.
fov_max_error=1, causes solve to fail completely!
But if I change the fov_estimate to 10.6, it runs fast.

So we are back to a confusion on fov. The RPi images are 14.3x10.6. The cedar-solver is reporting fov as 10.6. So it really seems the smaller dimension is used?

On the Pi5, excluding disk read and image conversion. I'm getting detect in about 7ms and solve in 5-10ms.
On the Pi zero, same conditions, detect is 40ms, solve in 40-90ms.

These are excellent results. Thank you again.

I didnt get any full frame images last night, UK summers generally are very high moisture so even 'clear' is often unusable. But I guess the conversion to 'L' is going to take even longer and might ruin everything.

Dale and I have kept in contact over the years as we have worked on similar challenges. I've also been studying PiFinder in some detail. As it should be, the final application is key. My own is to help position large and heavy Dobsonian telescopes to sub-arc minute accuracy.

**Reply by smroid** [reply]
Your Pi5 and Pi zero execution times are great, in line with my experience.

Regarding the FOV horizontal vs vertical puzzle: check your args to solve_from_centroids. The star_centroids arg is a list of (y, x) pairs (not x, y), and the size arg is a (height, width) tuple (not width, height).

As you've found, the HQ camera is not a great fit for Cedar, compared to the ASI120mm mini which is fantastic. But the HQ camera is less expensive, so it has that advantage (and it does work, well enough).

Glad it's working out for you! Keep me posted on your progress, and feel free to drop in on our [Discord](https://discord.gg/NmBmneQY) and/or post on our [CN thread](https://www.cloudynights.com/topic/927634-introducing-the-cedar-e-finder).

**Reply by smroid** [reply]
> Dale and I have kept in contact over the years as we have worked on similar challenges. I've also been studying PiFinder in some detail. As it should be, the final application is key. My own is to help position large and heavy Dobsonian telescopes to sub-arc minute accuracy.

Yes, I think we're all chasing the same goal, bringing joy to visual observing with any telescope.

I'm glad that Cedar is helping your application. If you feel so inclined, I would be happy if you could give public credit to Cedar's contributions to your system.

Also, just a quick note that cedar-detect is licensed for non-commercial use. Anything you build or other hobbyists build based on your descriptions of eFinder are OK, as we very much want to contribute to and support the DIY/hobby/hacker/amateur/enthusiast/educational community.

Clear skies!

**Comment by AstroKeith** [comment]
Ahah!
I need to put my numpy hat on.
I just entered (img.size) for the size arg.
Thanks again.

**Comment by AstroKeith** [comment]
Getting a new error after a fresh install. My install script is the same so I'm suspecting something has changed in the repo?
Error is
WARNING:root:RPC failed with: failed to decode Protobuf message: CentroidsRequest.sigma: invalid wire type: SixtyFourBit (expected ThirtyTwoBit)
ERROR:root:RPC failed with: failed to decode Protobuf message: CentroidsRequest.sigma: invalid wire type: SixtyFourBit (expected ThirtyTwoBit)
Any ideas
TIA
Keith

**Comment by smroid** [comment]
Yes. Do another pull on all repos, and re compile all proto files.

**Reply by smroid** [reply]
Let me know if this got you sorted out, and sorry for the breakage. I'm re-learning the age-old lesson to favor double precision floats. If you want to represent right ascension to arc sec precision, single precision float is on the hairy edge of being unsuitable.

BTW, I'm looking at your efinder-lite page, where you describe using Cedar-detect (thanks!) with Tetra3. You're using Cedar-solve as well as -detect, right? Cedar-solve provides much more consistent performance than Tetra3 (fewer failed or lengthy solves), so it might be worth mentioning that eFinder is using Cedar-detect and Cedar-solve (which is an evolution of Tetra3). And if you've not migrated from Tetra3 to Cedar-solve, you should :-). Just be sure to use the Cedar-solve default database, or generate a new one using Cedar-solve.

Thanks!

Relatedly: in my Cedar-solve and Cedar-detect pages, may I cite your eFinder project as a user?

**Comment by AstroKeith** [comment]
All sorted thanks.
I've changed my description as suggested on my repo. I'll update my website asap. The call to solve is the same as base tetra3 so I had assumed it was original tetra3 still.
I'm still using tetra3 for centroid extraction for my focus utility, at least for the moment, as I find it more stable with out of focus stars or overloads. For general use cedar is indeed proving very effective. It runs quite fast enough on the Pi Zero 2 for my application, which is essentially just a real sky align/sync for Nexus DSC based systems.

Please feel free to reference my project. Probably the link to eFinder-Lite on my website is best?
